File: 44. 通配符匹配/动态规划.swift
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        var target = Array(s), regex = [Character](), flag = false
        // 压缩 p 中多余的 "*"
        for c in p {
            if !(c == "*" && flag) {
                regex.append(c)
            }
            flag = c == "*"
        }
        // 访问表和记忆表
        var visit = [[Bool]](repeating: [Bool](repeating: false, count: target.count), count: regex.count)
        var table = [[Bool]](repeating: [Bool](repeating: false, count: target.count), count: regex.count)
        // 记忆函数，同时处理边界问题
        func memory(_ x: Int, _ y: Int) -> Bool {
            if x == -1 && y == -1 {
                return true
            } else if x == -1 {
                return false
            } else if y == -1 {
                return x == 0 && regex[0] == "*"
            } else {
                if !visit[x][y] {
                    table[x][y] = search(x, y)
                    visit[x][y] = true
                }
                return table[x][y]
            }
        }
        // 查找函数，放置动态转移方程
        func search(_ x: Int, _ y: Int) -> Bool {
            if regex[x] == "?" || regex[x] == target[y] {
                return memory(x - 1, y - 1)
            } else if regex[x] == "*" {
                return memory(x - 1, y) || memory(x, y - 1)
            } else {
                return false
            }
        }
        return memory(regex.count - 1, target.count - 1)
    }
}

File: 0494. 目标和/动态规划.js
// 1000ms
var findTargetSumWays = function(nums, target) {
    const len = nums.length
    let table = new Array(len).fill(0).map(i =>
        // -20000 ~ 20000
        new Array(40001).fill(-1)
    )

    function memory(i, sum) {
        if (i === -1) {
            return sum === 0 ? 1 : 0
        } else if (i >= 0) {
            if (table[i][sum + 20000] === -1) {
                table[i][sum + 20000] = search(i, sum)
            }
            return table[i][sum + 20000]
        } else {
            return 0
        }
    }

    function search(i, sum) {
        return memory(i - 1, sum + nums[i]) + memory(i - 1, sum - nums[i])
    }

    return memory(len - 1, target)
};

// dp[i][s] = dp[i-1][s+n[i]] + dp[i-1][s-n[i]]